src/query/graphite/native/aggregation_functions.go

```diff
@@ -278,6 +277,82 @@ func combineSeriesWithWildcards(
 	return r, nil
 }
 
+/*
+applyByNode takes a seriesList and applies some complicated function (described by a string), replacing templates with unique
+prefixes of keys from the seriesList (the key is all nodes up to the index given as `nodeNum`).
+
+If the `newName` parameter is provided, the name of the resulting series will be given by that parameter, with any
+"%" characters replaced by the unique prefix.
+
+Example:
+
+.. code-block:: none
+
+&target=applyByNode(servers.*.disk.bytes_free,1,"divideSeries(%.disk.bytes_free,sumSeries(%.disk.bytes_*))")
+
+Would find all series which match `servers.*.disk.bytes_free`, then trim them down to unique series up to the node
+given by nodeNum, then fill them into the template function provided (replacing % by the prefixes).
+
+Additional Examples:
+
+Given keys of
+
+- `stats.counts.haproxy.web.2XX`
+- `stats.counts.haproxy.web.3XX`
+- `stats.counts.haproxy.web.5XX`
+- `stats.counts.haproxy.microservice.2XX`
+- `stats.counts.haproxy.microservice.3XX`
+- `stats.counts.haproxy.microservice.5XX`
+
+The following will return the rate of 5XX's per service:
+
+.. code-block:: none
+
+applyByNode(stats.counts.haproxy.*.*XX, 3, "asPercent(%.5XX, sumSeries(%.*XX))", "%.pct_5XX")
+
+The output series would have keys `stats.counts.haproxy.web.pct_5XX` and `stats.counts.haproxy.microservice.pct_5XX`.
+*/
+func applyByNode(ctx *common.Context, seriesList singlePathSpec, nodeNum int, templateFunction string, newName string) (ts.SeriesList, error) {
+	// using this as a set
+	prefixMap := make(map[string]int)
+	for _, series := range seriesList.Values {
+		prefix := strings.Join(strings.Split(series.Name(), ".")[:nodeNum+1], ".")
+		prefixMap[prefix] = 1
+	}
```

Comment:
This will panic if `nodeNum` is greater than the number of `.` in the name. Instead, can probably do something like this (which ends up being more efficient too)

```golang
for _, series := range seriesList.Values {
		var (
			name  = series.Name()

			partsSeen int
			prefix    string
		)

		for i, c := range name {
			if c == '.' {
				partsSeen++
				if partsSeen == nodeNum+1 {
					prefix = name[:i]
					break
				}
			}
		}

		if len(prefix) == 0 {
			continue
		}

		prefixMap[prefix] = struct{}{}
}
```

Reply:
I integrated this logic. Hopefully you're happy with the way it is now.